src/app/server/CommissioningWindowManager.cpp

```diff
@@ -0,0 +1,276 @@
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/server/CommissioningWindowManager.h>
+#include <app/server/Mdns.h>
+#include <app/server/Server.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/SafeInt.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <transport/SecureSessionMgr.h>
+
+#if CHIP_ENABLE_OPENTHREAD
+#include <platform/ThreadStackManager.h>
+#endif
+#include <lib/mdns/Advertiser.h>
+
+namespace {
+
+void HandleCommissioningWindowTimeout(chip::System::Layer * aSystemLayer, void * aAppState)
+{
+    chip::CommissioningWindowManager * commissionMgr = static_cast<chip::CommissioningWindowManager *>(aAppState);
+    commissionMgr->CloseCommissioningWindow();
+}
+
+void OnPlatformEventWrapper(const chip::DeviceLayer::ChipDeviceEvent * event, intptr_t arg)
+{
+    chip::CommissioningWindowManager * commissionMgr = reinterpret_cast<chip::CommissioningWindowManager *>(arg);
+    commissionMgr->OnPlatformEvent(event);
+}
+} // namespace
+
+namespace chip {
+
+void CommissioningWindowManager::OnPlatformEvent(const DeviceLayer::ChipDeviceEvent * event)
+{
+    if (event->Type == DeviceLayer::DeviceEventType::kCommissioningComplete)
+    {
+        if (event->CommissioningComplete.status == CHIP_NO_ERROR)
+        {
+            ChipLogProgress(Discovery, "Commissioning completed successfully");
+        }
+        else
+        {
+            ChipLogError(Discovery, "Commissioning errored out with error %" CHIP_ERROR_FORMAT,
+                         event->CommissioningComplete.status.Format());
+        }
+        // reset all advertising
+        app::MdnsServer::Instance().StartServer(Mdns::CommissioningMode::kDisabled);
+    }
+    else if (event->Type == DeviceLayer::DeviceEventType::kOperationalNetworkEnabled)
+    {
+        app::MdnsServer::Instance().AdvertiseOperational();
+        ChipLogError(Discovery, "Operational advertising enabled");
+    }
+}
+
+void CommissioningWindowManager::Cleanup()
+{
+    mServer->GetExchangManager().UnregisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::PBKDFParamRequest);
+    StopAdvertisement();
+
+    // reset all advertising
+    app::MdnsServer::Instance().StartServer(Mdns::CommissioningMode::kDisabled);
+}
+
+void CommissioningWindowManager::OnSessionEstablishmentError(CHIP_ERROR err)
+{
+    Cleanup();
+
+    ChipLogProgress(AppServer, "OnSessionEstablishmentError: %s", ErrorStr(err));
+    ChipLogProgress(AppServer, "Failed in SPAKE2+ handshake");
```

Comment:
Is this log always true? The OnSessionEstablishmentError seems to be distributed in many places and I  wonder if some have not even started the handshake (e.g. running out of memory when allocating a pairing would also trigger it)

Reply:
In this class, `OnSessionEstablishmentError` will only be called from `PASESession`, as that's where the callback/delegate is being registered.

Reply:
My worry is to not send people on a bad path if the error message makes assumption on wha the code was doing at the time. The message 'You got an error: errstr' is generic, however saying a specific state seemed to make assumptions. If  this is wrong feel free to just resolve the comment.

I still wonder if this should be ChipLogError though - this only happens on error.

Reply:
Ok, I understand your concerns. I'll update the log to error, and update the message.